### PR TITLE
Added ability to add authentication logic into pushgateway request

### PR DESCRIPTION
**Pull request by bitsal**
Added Basic Auth. implementation.

@brian-brazil 
See https://github.com/prometheus/client_java/issues/364.

### CODE REVIEW

**Comment by brian-brazil** [comment]
This only supports basic auth. I'd prefer to make this pluggable so any auth method can be used.

**Comment by bitsal** [comment]
@brian-brazil 
1. Now you don't have any auth
2. I've implemented only one auth mechanism. Anyone else will be able to implement own even outside of the library (I've left a door for this)

If you have any code suggestions/critics just write it down I'm opened to implement it of course if it is not so much.

>I'd prefer to make this pluggable so any auth method can be used
Probably, I don't understand what exactly you mean. Add more details about your idea(s).

**Comment by brian-brazil** [comment]
How would this work if someone wanted say TLS client cert auth, or to adjust TLS settings?

I'd like to avoid the situation where I have to maintain every possible http configuration.

**Comment by bitsal** [comment]
Easy. It is necessary to create a class like `TlsPushGateway` and override `authenticate()` based on this e.g. https://stackoverflow.com/questions/30121510/java-httpsurlconnection-and-tls-1-2
```
protected void authenticate(HttpURLConnection connection) {
     SSLContext sc = SSLContext.getInstance("TLSv1");
     // Init the SSLContext with a TrustManager[] and SecureRandom()
     sc.init(null, trustCerts, new java.security.SecureRandom());
     
     ((HttpsURLConnection) connection).setSSLSocketFactory(sc.getSocketFactory());
}
```

**Comment by amlozano** [comment]
I think that a better solution would be to extract the code from the Http client outside PushGateway code, and inject it into the constructor. In that case you can compose PushGateway with the HttpClient that you want, allowing any kind of extension open for the user.

**Comment by bitsal** [comment]
@amlozano Agree. A composition is much better. I've tried not to add a lot of code because I don't know what limitations are of using other libraries for http(s) requests.

@brian-brazil Could you tell me about such limitations? Like Java version, dependencies and so on.

**Comment by amlozano** [comment]
@bitsal What I meant is to extract the line of 
`HttpURLConnection connection = (HttpURLConnection) new URL(url).openConnection();`
The class PushGateway depends on the abstract class HttpURLConnection, so the creation of the exact implementation is outside the code. In that way you can create an HttpsURLConnection configuring the 
`connection.setRequestProperty("Authorization", basicAuthHeader);`
or the 
`connection.setSSLSocketFactory(sc.getSocketFactory());`
outside PushGateway.

**Comment by brian-brazil** [comment]
The only real consideration is that we need to stick with Java 6 features.

**Comment by bitsal** [comment]
@amlozano It was clear, thanks. I will add composition and try to be back-compatible just not to break existing code.

File: simpleclient_pushgateway/src/main/java/io/prometheus/client/exporter/BasicAuthHttpConnectionFactory.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    public HttpURLConnection create(String url) throws IOException {
+        HttpURLConnection connection = originConnectionFactory.create(url);
+        connection.setRequestProperty("Authorization", basicAuthHeader);
+        return connection;
+    }
+
+    private String encode(String user, String password) {
+        try {
+            byte[] message = (user + ":" + password).getBytes("UTF-8");
+            String encoded = javax.xml.bind.DatatypeConverter.printBase64Binary(message);
```
Comment (brian-brazil) [inline simpleclient_pushgateway/src/main/java/io/prometheus/client/exporter/BasicAuthHttpConnectionFactory.java]: This should be imported.

Comment (bitsal) [reply simpleclient_pushgateway/src/main/java/io/prometheus/client/exporter/BasicAuthHttpConnectionFactory.java]: Fixed

**Comment by brian-brazil** [comment]
Can you add the DCO and update the docs please?

**Comment by bitsal** [comment]
Updated docs and added sign-off commit.

**Comment by bitsal** [comment]
@brian-brazil DCO is still red. I wonder how it works.
I've installed it using 
```
gem install dco
dco enable
```

File: README.md [hunk 2]
```diff
@@ -568,12 +568,34 @@ void executeBatchJob() throws Exception {
 }
  ```
 
-A separate registry is used, as the default registry may contain other metrics
-such as those from the Process Collector. See the 
-[Pushgateway documentation](https://github.com/prometheus/pushgateway/blob/master/README.md)
-for more information.
+ A separate registry is used, as the default registry may contain other metrics
```
Comment (brian-brazil) [inline README.md]: No leading spaces on these lines

**Comment by amlozano** [comment]
@bitsal I still see it better to be in a constructor rather than in the setConnectionFactory. Otherwise you will create something that later on has to be replaced. I.E. 
`myGateway = new PushGateway(url);
myGateway.setConnectionFactory(factory);`

the second line is replacing the URL configuration, which is mandatory for no reason.
I see it more natural to define a new constructor with the HttpConnectionFactory, so you can use either the address, the URL or the ConnectionFactory to create the connection.

BTW, I am not part of the project, but we use it in our team, and it would be nice as an user to have that in place.

File: README.md (hunk 2 above)
Comment (bitsal) [reply README.md]: Removed

**Comment by bitsal** [comment]
@amlozano `BasicAuthHttpConnectionFactory` should add only basic auth based on single responsibility principle if you want to configure HTTPS you should add own `HttpsConnectionFactory` that will decorate `BasicAuthHttpConnectionFactory` and add HTTPS logic.
```java
new BasicAuthHttpConnectionFactory(new HttpsConnectionFactory(), 'username', 'verySecretPassword');
```

**Comment by bitsal** [comment]
@brian-brazil Probably I should remove `BasicAuthPushGateway` beacuse it is more as an example than real helpful gateway. What do you think?

**Comment by amlozano** [comment]
@bitsal I agree on your comment, I just said that the BasicAuthPushGateway is enforcing you to use an HTTP connection, since it uses the constructor from PushGateway that sets an http url.

I think that class doesn't need to exist, since it is already easy to configure the BasicAuth just with composition.

**Comment by brian-brazil** [comment]
Yeah we can simplify, as long as it's covered in the docs.

**Comment by bitsal** [comment]
Done

**Comment by brian-brazil** [comment]
That all looks good, can you ensure all commits have the DCO?

**Comment by bitsal** [comment]
@brian-brazil not all of course, first ones are not
Is there a way to add to these first commits DCO?

**Comment by brian-brazil** [comment]
The easiest way is to squash your commits.

**Comment by bitsal** [comment]
I have never done it before. @brian-brazil do you have an exact command(s) how to do it without recreation a branch?

**Comment by brian-brazil** [comment]
`git rebase -i master` and set all but one commit to squash.

**Comment by bitsal** [comment]
@brian-brazil create a new PR https://github.com/prometheus/client_java/pull/398

**Pull request closed without merging**